lib/northbound_notif.c

```diff
@@ -0,0 +1,156 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+/*
+ * December 1 2024, Christian Hopps <chopps@labn.net>
+ *
+ * Copyright (c) 2024, LabN Consulting, L.L.C.
+ *
+ */
+#include <zebra.h>
+#include "debug.h"
+#include "typesafe.h"
+#include "northbound.h"
+
+#define __dbg(fmt, ...)	    DEBUGD(&nb_dbg_notif, "NB_OP_CHANGE: %s: " fmt, __func__, ##__VA_ARGS__)
+#define __log_err(fmt, ...) zlog_err("NB_OP_CHANGE: %s: ERROR: " fmt, __func__, ##__VA_ARGS__)
+
+static void nb_notif_add(const char *path)
+{
+}
+
+
+static void nb_notif_delete(const char *path)
+{
+}
+
+struct lyd_node *nb_op_update(struct lyd_node *tree, const char *path, const char *value)
+{
+	struct lyd_node *dnode;
+	const char *abs_path = NULL;
+
+
+	__dbg("updating path: %s with value: %s", path, value);
+
+	dnode = yang_state_new(tree, path, value);
+
+	if (path[0] == '/')
+		abs_path = path;
+	else {
+		abs_path = lyd_path(dnode, LYD_PATH_STD, NULL, 0);
+	}
+
+	nb_notif_add(abs_path);
+
+	if (abs_path != path)
```

Comment:
What is the chance that abs_path and path are equal and path was passed in?

Reply:
`abs_path` starts out assigned as `NULL`, it is then either assigned the value `path` or it is a new allocation from `lyd_path()`. `path` is not currently allowed to be NULL (notice we use `path[0]` without checking for `path == NULL`). So path can only ever equal `abs_path` when we set it that way (which is what our equality check is deciding, did we set it or newly allocate it).

Reply:
That said, I don't see a reason we couldn't also support a NULL path for updating the path/value of the tree node itself, so I'll modify the code to also handle path == NULL.

Reply:
actually nm, this gets tricky deeper in the implementation, i'll leave extended functionality for later if we need it.